File: framework/web/src/main/java/cn/bestwu/simpleframework/support/client/ClientHttpRequestWrapper.java
package cn.bestwu.simpleframework.support.client;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.net.URI;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.client.ClientHttpRequest;
import org.springframework.http.client.ClientHttpResponse;

/**
 * @author Peter Wu
 */
public class ClientHttpRequestWrapper implements ClientHttpRequest {

  private Logger log = LoggerFactory.getLogger(ClientHttpRequestWrapper.class);
  private final ClientHttpRequest request;
  private final ByteArrayOutputStream record = new ByteArrayOutputStream();

  public ClientHttpRequestWrapper(ClientHttpRequest request) {
    this.request = request;
  }

  @Override
  public ClientHttpResponse execute() throws IOException {
    ClientHttpResponse response = request.execute();
    if (log.isDebugEnabled()) {
      log.debug("request heads:{}",request.getHeaders());
      log.debug("request:{}", record.toString());
    }
    return response;
  }

  @Override
  public HttpMethod getMethod() {
    return request.getMethod();
  }

  @Override
  public String getMethodValue() {
    return request.getMethodValue();
  }

  @Override
  public URI getURI() {
    return request.getURI();
  }

  @Override
  public HttpHeaders getHeaders() {
    return request.getHeaders();
  }

  @Override
  public OutputStream getBody() throws IOException {
    return new OutputStreamWrapper(request.getBody());
  }

  private class OutputStreamWrapper extends OutputStream {

    private final OutputStream delegate;


    private OutputStreamWrapper(OutputStream delegate) {
      this.delegate = delegate;
    }

    @Override
    public void write(int b) throws IOException {
      delegate.write(b);
      record.write(b);
    }
  }
}

File: framework/starter-logging/src/main/kotlin/cn/bestwu/logging/client/ClientHttpRequestWrapper.kt
package cn.bestwu.logging.client

import cn.bestwu.logging.RequestLoggingConfig
import cn.bestwu.logging.client.ClientHttpRequestWrapper
import cn.bestwu.logging.operation.Operation
import cn.bestwu.logging.operation.RequestConverter
import cn.bestwu.logging.operation.ResponseConverter
import org.slf4j.LoggerFactory
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpMethod
import org.springframework.http.client.ClientHttpRequest
import org.springframework.http.client.ClientHttpResponse
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.OutputStream
import java.net.URI
import java.time.LocalDateTime

/**
 * @author Peter Wu
 */
class ClientHttpRequestWrapper(
    private val collectionName: String,
    private val name: String,
    private val request: ClientHttpRequest
) : ClientHttpRequest {
    private val log = LoggerFactory.getLogger(ClientHttpRequestWrapper::class.java)
    val record = ByteArrayOutputStream()

    @Throws(IOException::class)
    override fun execute(): ClientHttpResponse {
        val dateTime =
            if (log.isInfoEnabled) {
                LocalDateTime.now()
            } else null
        var response = request.execute()
        if (log.isInfoEnabled) {
            response = ClientHttpResponseWrapper(response)
            val operation = Operation(
                collectionName = collectionName,
                name = name,
                protocol = "HTTP/1.1",
                request = RequestConverter.convert(this, dateTime!!),
                response = ResponseConverter.convert(response)
            )
            log.info(
                operation.toString(
                    RequestLoggingConfig(
                        true,
                        true,
                        true,
                        arrayOf(),
                        arrayOf(),
                        true,
                        true,
                        -1
                    )
                )
            )
        }
        return response
    }

    override fun getMethod(): HttpMethod {
        return request.method!!
    }

    override fun getMethodValue(): String {
        return request.methodValue
    }

    override fun getURI(): URI {
        return request.uri
    }

    override fun getHeaders(): HttpHeaders {
        return request.headers
    }

    @Throws(IOException::class)
    override fun getBody(): OutputStream {
        return OutputStreamWrapper(request.body)
    }

    private inner class OutputStreamWrapper constructor(private val delegate: OutputStream) :
        OutputStream() {
        @Throws(IOException::class)
        override fun write(b: Int) {
            delegate.write(b)
            record.write(b)
        }
    }
}